### ISSUE TITLE
Manage taxonomies properties in an external CMS

### ISSUE THREAD

**Issue by raphael0202** [openfoodfacts/openfoodfacts-server#8536]
This issue arose from recent discussions we had with @alexgarel and @teolemon and the rest of the permanent team about the way we manage taxonomy-related content on Open Food Facts.

Currently, everything is stored in the taxonomy files: descriptions, wikidata IDs, vegan/vegetarian status for ingredients,...
This approach has served us well so far, but has some drawbacks:
- taxonomy syntax is error-prone (as any syntax)
- the contribution barrier is high (mastering git(hub), having to do a pull request even for a typo)
- it makes taxonomy files bloated
- it's hard to scale property translations (see #8067)

We *do* need careful review when updating taxonomy names, as it has a major impact on ingredient parsing and it can be quite tricky. But we probably don't when it comes to fixing a typo, adding a basic description or updating a wikidata ID: user tracking is enough to spot potential abuse.

My proposal is to use an external CMS to manage taxonomy properties. I've tried [strapi](https://strapi.io/) which is very flexible headless CMS. This can be used to update entries with custom schemas.

Schema management:
![Capture d’écran du 2023-06-08 16-11-33](https://github.com/openfoodfacts/openfoodfacts-server/assets/9609923/6b5ae57b-0d5e-40dd-9e11-3a402d061e42)

Items list:
![Capture d’écran du 2023-06-08 16-12-03](https://github.com/openfoodfacts/openfoodfacts-server/assets/9609923/fe2ea45d-7a5b-48c2-82a3-0d5f1c9acea2)

Item update:
![Capture d’écran du 2023-06-08 16-12-23](https://github.com/openfoodfacts/openfoodfacts-server/assets/9609923/81e2bc2a-aae7-4c47-9ad1-a0c899213551)

The tool is open-source and allow easy import/export of data through an API, we could export it to JSON/yaml format. We could perform a daily export of the data so that Product Opener (and other projects) can access it. The advantages of this approach are the following:

- allow easier contribution (web contribution)
- no need to develop in-house tool, easy to set-up
- flexible schema management: it's possible to add new properties (we can also add limited constraints to user input)
- i18n support
- everything is callable through API, which allows easy ChatGPT/ML integrations

After talking with @alexgarel, it turns out it would be better to manage synonyms and names through taxonomy editor only, as we want to control user input sanitization (checking synonyms/duplicates).

edit: To give a bit more context, we would like to add more "knowledge" information to Open Food Facts, such as the claims associated with food labels, health risks associated with additives, category descriptions,... It turned out quickly that we cannot reasonably store all this information in taxonomy files, hence this proposal.

**Comment by john-gom** [comment]
It is something I have thought about a lot too since @teolemon mentioned the idea of using Wikibase / Wikidata. There are certainly advantages to a generic tool but we do need extensibility for things like constraint validation and concepts like synonyms (which Wikibase does handle). Another area where I think taxonomy management could be enhanced is being able to see the impact of changes, e.g. "if I update the nutrient content of this ingredient how many products will be affected?".

In my own experience these generic tools never go quite far enough and then when you try to customise them you get into a world of pain with proprietary techniques, no testability, etc. and so it would be easier to develop from scratch. I would personally go further than the current taxonomy editor in that I would want specific UIs for each type of taxonomy. We can easily cater for extensible attributes and I have done this many times before.

One thing I 100% agree on is that we should treat taxonomies as data and not source code. Unit tests should rely on specifically created sample taxonomies for the purpose of the specific test, with validation rules during taxonomy data capture to ensure that edits to the data don't break things.

**Comment by raphael0202** [comment]
I agree with you on the customization part. The issue is that taxonomy editor is not ready for deployment, and we lack development resources on both front-end and backend.
I doubt we could have a version of taxonomy editor (for property editing) that has the same features as an external CMS for content creation in 2023, and OFF knowledge base is a project I really want to launch this year.
For purely text-based content without validation, a CMS is suited, but synonyms or properties like nutrient content it would be better to keep it on taxonomy editor.

I've thought on relying on Wikidata for this content, but as far as I recall we cannot create custom properties anymore on wikidata.

**Comment by john-gom** [comment]
I'm not sure that having two tools for this (CMS + Taxonomy editor) will be very usable. I take the point about not having resources, and it is disappointing that existing tools don't quite do the job, but I suspect that if we try and use a CMS to do part of the work and then have other tools to supplement data, do validation and export in various formats, then we could end up doing a lot more work in the long term than just writing the tool from scratch.

Having said that, it might be possible that we could start with a CMS like Strapi and write plug-ins to handle specific things like Synonyms and other business rules?

Maybe we should start by simply capturing all of the requirements, including what @stephanegigandet mentioned above and then think about the various options?

**Comment by VaiTon** [comment]
I have recently found https://terminusdb.com/ and, despite not being as mature as something like Wikibase, it has an integrated review and versioning workflow.

**Comment by john-gom** [comment]
> I have recently found https://terminusdb.com/ and, despite not being as mature as something like Wikibase, it has an integrated review and versioning workflow.

Did you find any information on how it supports multiple translations of entry names and descriptions? I couldn't see anything on a quick scan of the docs.

**Comment by alexgarel** [comment]
Another path might be using something like a no-code tool (like airtable). 
I think about [nocodb](https://www.nocodb.com/) or [baserow](https://baserow.io/)
I've also seen [data-patch](https://www.multi.coop/products?locale=fr&item=.%2Ftexts%2Freferences%2Fprojects%2Fdata-patch.md) but far less mature. The advantage here is that the backend is postgresql. So we could still do a lot of checks / stats.

**Comment by raphael0202** [comment]
I'm closing this issue, as taxonomy editor is now usable, and it makes more sense to add support for this feature in taxonomy editor.
